File: clients/Ruby-client/lib/solidb/columnar_client.rb
module SoliDB
  class ColumnarClient
    def initialize(client)
      @client = client
    end

    def create(name, columns)
      @client.send_command("create_columnar_table", database: @client.database, name: name, columns: columns)
    end

    def list
      @client.send_command("list_columnar_tables", database: @client.database) || []
    end

    def get(name)
      @client.send_command("get_columnar_table", database: @client.database, name: name)
    end

    def delete(name)
      @client.send_command("delete_columnar_table", database: @client.database, name: name)
      nil
    end

    def insert(name, rows)
      @client.send_command("columnar_insert", database: @client.database, name: name, rows: rows)
    end

    def query(name, query, params: nil)
      args = { database: @client.database, name: name, query: query }
      args[:params] = params if params
      @client.send_command("columnar_query", args) || []
    end

    def aggregate(name, aggregation)
      @client.send_command("columnar_aggregate", database: @client.database, name: name, aggregation: aggregation)
    end

    def create_index(table_name, index_name:, column:, index_type: nil)
      args = {
        database: @client.database,
        table_name: table_name,
        index_name: index_name,
        column: column
      }
      args[:index_type] = index_type if index_type
      @client.send_command("columnar_create_index", args)
    end

    def list_indexes(table_name)
      @client.send_command("columnar_list_indexes", database: @client.database, table_name: table_name) || []
    end

    def delete_index(table_name, index_name)
      @client.send_command("columnar_delete_index", database: @client.database, table_name: table_name, index_name: index_name)
      nil
    end

    def add_column(table_name, column_name:, column_type:, default_value: nil)
      args = {
        database: @client.database,
        table_name: table_name,
        column_name: column_name,
        column_type: column_type
      }
      args[:default_value] = default_value if default_value
      @client.send_command("columnar_add_column", args)
      nil
    end

    def drop_column(table_name, column_name)
      @client.send_command("columnar_drop_column", database: @client.database, table_name: table_name, column_name: column_name)
      nil
    end

    def stats(table_name)
      @client.send_command("columnar_stats", database: @client.database, table_name: table_name)
    end
  end
end

File: clients/PHP-client/src/ColumnarClient.php
<?php

namespace SoliDB;

class ColumnarClient
{
    private Client $client;

    public function __construct(Client $client)
    {
        $this->client = $client;
    }

    public function create(string $name, array $columns): array
    {
        $res = $this->client->sendCommand('create_columnar_table', [
            'database' => $this->client->getDatabase(),
            'name' => $name,
            'columns' => $columns
        ]);
        return $res['data'] ?? [];
    }

    public function list(): array
    {
        $res = $this->client->sendCommand('list_columnar_tables', ['database' => $this->client->getDatabase()]);
        return $res['data'] ?? [];
    }

    public function get(string $name): array
    {
        $res = $this->client->sendCommand('get_columnar_table', [
            'database' => $this->client->getDatabase(),
            'name' => $name
        ]);
        return $res['data'] ?? [];
    }

    public function delete(string $name): void
    {
        $this->client->sendCommand('delete_columnar_table', [
            'database' => $this->client->getDatabase(),
            'name' => $name
        ]);
    }

    public function insert(string $name, array $rows): array
    {
        $res = $this->client->sendCommand('columnar_insert', [
            'database' => $this->client->getDatabase(),
            'name' => $name,
            'rows' => $rows
        ]);
        return $res['data'] ?? [];
    }

    public function query(string $name, string $query, ?array $params = null): array
    {
        $args = [
            'database' => $this->client->getDatabase(),
            'name' => $name,
            'query' => $query
        ];
        if ($params !== null) {
            $args['params'] = $params;
        }
        $res = $this->client->sendCommand('columnar_query', $args);
        return $res['data'] ?? [];
    }

    public function aggregate(string $name, array $aggregation): array
    {
        $res = $this->client->sendCommand('columnar_aggregate', [
            'database' => $this->client->getDatabase(),
            'name' => $name,
            'aggregation' => $aggregation
        ]);
        return $res['data'] ?? [];
    }

    public function createIndex(string $tableName, string $indexName, string $column, ?string $indexType = null): array
    {
        $args = [
            'database' => $this->client->getDatabase(),
            'table_name' => $tableName,
            'index_name' => $indexName,
            'column' => $column
        ];
        if ($indexType !== null) {
            $args['index_type'] = $indexType;
        }
        $res = $this->client->sendCommand('columnar_create_index', $args);
        return $res['data'] ?? [];
    }

    public function listIndexes(string $tableName): array
    {
        $res = $this->client->sendCommand('columnar_list_indexes', [
            'database' => $this->client->getDatabase(),
            'table_name' => $tableName
        ]);
        return $res['data'] ?? [];
    }

    public function deleteIndex(string $tableName, string $indexName): void
    {
        $this->client->sendCommand('columnar_delete_index', [
            'database' => $this->client->getDatabase(),
            'table_name' => $tableName,
            'index_name' => $indexName
        ]);
    }

    public function addColumn(string $tableName, string $columnName, string $columnType, mixed $defaultValue = null): void
    {
        $args = [
            'database' => $this->client->getDatabase(),
            'table_name' => $tableName,
            'column_name' => $columnName,
            'column_type' => $columnType
        ];
        if ($defaultValue !== null) {
            $args['default_value'] = $defaultValue;
        }
        $this->client->sendCommand('columnar_add_column', $args);
    }

    public function dropColumn(string $tableName, string $columnName): void
    {
        $this->client->sendCommand('columnar_drop_column', [
            'database' => $this->client->getDatabase(),
            'table_name' => $tableName,
            'column_name' => $columnName
        ]);
    }

    public function stats(string $tableName): array
    {
        $res = $this->client->sendCommand('columnar_stats', [
            'database' => $this->client->getDatabase(),
            'table_name' => $tableName
        ]);
        return $res['data'] ?? [];
    }
}
